intellij/src/saros/intellij/ui/views/buttons/AbstractSessionToolbarButton.java

```diff
@@ -62,6 +83,21 @@ void setInitialState() {
     }
   }
 
+  @Override
+  public final void dispose() {
+    sarosSessionManager.removeSessionLifecycleListener(sessionLifecycleListener);
+
+    disposeComponents();
+  }
+
+  /**
+   * Method to dispose components of the button.
+   *
+   * <p>This method is called when the project the button belongs to is disposed. It should drop any
+   * internal state that would prevent the object from being garbage collected.
+   */
+  abstract void disposeComponents();
```

Comment:
protected abstract void disposeComponents()

Reply:
Hmmm...*Technically*, this is currently not needed as the class is package-private and all its implementations are contained in the package. (This also means that I should make the instance variables package private as well.)

But this still creates a potential problem with the implementing classes, which are not package private. If I were to create another class outside the package that extends such an implementation (e.g. the `LeaveSessionButton`), I would no longer be able to override these methods.

But I am not sure whether this is something to worry about as it (currently) makes sense to keep all implementations in this package.

Reply:
But is it really worth it to change the visibility? If this is done for this method, it should also subsequently be done for all package-private methods and variables in this class and `AbstractToolbarButton`. This could be done, but I don't really see the advantage of this as all implementations should most likely be located in this package. (As mentioned in my initial comment, this is only a problem when creating a class located outside the package that is extending an implementation.)

Furthermore, this also leads to IntelliJ complaining that the variables/methods could have a more strict access restrictions.
This in itself is not really a 'problem', but it it annoying when looking at the code (and can lead to reader suspecting issues where there are none). This can be avoided by adding annotations suppressing the warnings, but this has to be done for every instance and also adds clutter.

Reply:
Why is IntelliJ complaining about such an issue ? If I create a class with multiple protected modifiers (because the class should be subclassed somewhere in the future) it makes no sense to complain about that.

Of course it is fine for now to use package private but imagine you have

saros.*.ui.*.featureXYZ

now you add saros.*.ui.*.featureXYZ.subfeatureABC which should use classes in the parent package and now you are out of luck because the standard modifiers in Java are quite bad when it comes a more complex packet structure.

I guess this will change with Java 9, see https://openjdk.java.net/projects/jigsaw/

Reply:
Yeah, the access modifiers in java are somewhat lacking when it comes to inheritance or complex package structures.

> Why is IntelliJ complaining about such an issue ?

Maybe "complaining" was the wrong word. It gets marked as an issue by the code analyzer in the editor as the analyzer notices that the access modifier could be more strict *with the current usage*. This generally makes sense in my opinion to remind developers to encapsulate the class as  well es possible. However, as this is a static analysis, it of course can't deal with planned/future developments that would require a less strict access modifier.

As already mentioned, the rule can be disabled or it can be overridden through annotations, but I would still prefer to avoid it in this case as it would require a lot of annotations (5 out of the 8 fields/methods would have to be annotated).
In general, I try to resolve any issues that the code analyzer has (by following the suggestion where appropriate or adding an annotation to exclude the piece of code from the rule). This allows me to better notice when I have accidentally introduces a new issue while changing parts of the code.